woocommerce/woocommerce-gateway-stripe
# Prevent unnecessary prompt on page refresh

All changes in the Stripe settings page are saved on the `Save` button click, except for two new options which are introduced for non-UPE settings.
1. Changing the display order of the payment methods
2. Customizing individual payment methods

These two new features have their own `Save changes` button and an API request is made to the backend on clicking the `Save changes` button.

For both of them, after saving the data by clicking `Save changes` button, if you refresh the page you will see a browser prompt saying you have unsaved changes (which is wrong).

In this PR, I have addressed this issue. Now after saving the changed order or customizing any payment method you will no longer get the prompt on page refresh. (prompt will still appear if you have changed any other values except these two)


## Changes proposed in this Pull Request:
- Introduced [task/updated-settings-page](https://github.com/woocommerce/woocommerce-gateway-stripe/tree/task/updated-settings-page) `onSaveChanges` function which updates the `initialSettings` object which holds the already saved data. This `initialSettings` is compared with the `settings` object (which holds any unsaved changes) on page refresh.
- Changed the signature of `customizePaymentMethod` function to support the above changes.


## Testing instructions
- Go to the Stripe settings page. Disable UPE.
- Click `Change display order` and change the order of payment methods. Click `Save changes`.
- Refresh the page. You should not see any prompt.
- Click `Customize` beside any of the payment methods and change the name/description. Click `Save changes`.
- Refresh the page. You should not see any prompt.
- Change any other setting (i.e., enable/disable a method, update the bank statement field, etc.).
- Refresh the page. You should see the prompt about unsaved changes.

![prompt](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/7ab7c064-bcbf-4bef-9ae6-3e69dc98caa6)
